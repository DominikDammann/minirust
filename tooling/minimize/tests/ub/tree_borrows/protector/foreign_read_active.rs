//@ compile-flags: --minimize-tree-borrows

// Check that a foreign read of a protected mutable reference with active permission is UB.

fn foo(x: *mut i32, y: &mut i32) {
    // (x, Reserved) -> (y, P[Reserved])
    *y = 57; // (x, Active) -> (y, P[Active])
    unsafe { assert!(*x == 57) }; // UB! Foreign Read from Protected Active.
}

fn main() {
    let x = &mut 31; // (x, Reserved)   
    let xraw = x as *mut i32; // (x, Reserved) 
    let y = unsafe { &mut *xraw }; // (x, Reserved) -> (y, Reserved)

    foo(xraw, y);
}
