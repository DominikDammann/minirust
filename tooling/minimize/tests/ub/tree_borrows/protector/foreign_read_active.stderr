fatal error: UB: Tree Borrows: a protected pointer with Active permission becomes Disabled
